# Integrated Project 1

In this project, I will be using historical game sales data from 2016 to create a forecast for the following sales year by:<br>
- Exploring game releases and sales across different years, examining platform popularity and their periods of relevancy. 
<br>
- Identifying leading platforms, track sales trends, and pinpoint potentially profitable options for the company.
<br> 
- Analyzing the impact of reviews on sales for a chosen platform and compare game sales across platforms. 
- Investigating genre distribution to track trends in profitability.
<br>

### Data description
The dataset contains the following details:
- Name 

- Platform 

- Year_of_Release 

- Genre 

- NA_sales (North American sales in USD million) 

- EU_sales (sales in Europe in USD million) 

- JP_sales (sales in Japan in USD million) 

- ther_sales (sales in other countries in USD million) 

- Critic_Score (maximum of 100) 

- User_Score (maximum of 10) 

- Rating (ESRB)

This data may be incomplete and will be processed before analysis.

In [9]:
# importing 
import pandas as pd
import matplotlib as plt
import seaborn as sns
import plotly.express as px

# Read in the data and view for any abnormalities
games = pd.read_csv('/Users/angeneris/Desktop/integrated_project_1/games.csv')
games.head(20)

Name Platform  Year_of_Release  \
0                                     Wii Sports      Wii           2006.0   
1                              Super Mario Bros.      NES           1985.0   
2                                 Mario Kart Wii      Wii           2008.0   
3                              Wii Sports Resort      Wii           2009.0   
4                       Pokemon Red/Pokemon Blue       GB           1996.0   
5                                         Tetris       GB           1989.0   
6                          New Super Mario Bros.       DS           2006.0   
7                                       Wii Play      Wii           2006.0   
8                      New Super Mario Bros. Wii      Wii           2009.0   
9                                      Duck Hunt      NES           1984.0   
10                                    Nintendogs       DS           2005.0   
11                                 Mario Kart DS       DS           2005.0   
12                   Pokemon Gold/Pokemon Silver       GB           1999.0   
13                                       Wii Fit      Wii           2007.0   
14                            Kinect Adventures!     X360           2010.0   
15                                  Wii Fit Plus      Wii           2009.0   
16                            Grand Theft Auto V      PS3           2013.0   
17                 Grand Theft Auto: San Andreas      PS2           2004.0   
18                             Super Mario World     SNES           1990.0   
19  Brain Age: Train Your Brain in Minutes a Day       DS           2005.0   

           Genre  NA_sales  EU_sales  JP_sales  Other_sales  Critic_Score  \
0         Sports     41.36     28.96      3.77         8.45          76.0   
1       Platform     29.08      3.58      6.81         0.77           NaN   
2         Racing     15.68     12.76      3.79         3.29          82.0   
3         Sports     15.61     10.93      3.28         2.95          80.0   
4   Role-Playing     11.27      8.89     10.22         1.00           NaN   
5         Puzzle     23.20      2.26      4.22         0.58           NaN   
6       Platform     11.28      9.14      6.50         2.88          89.0   
7           Misc     13.96      9.18      2.93         2.84          58.0   
8       Platform     14.44      6.94      4.70         2.24          87.0   
9        Shooter     26.93      0.63      0.28         0.47           NaN   
10    Simulation      9.05     10.95      1.93         2.74           NaN   
11        Racing      9.71      7.47      4.13         1.90          91.0   
12  Role-Playing      9.00      6.18      7.20         0.71           NaN   
13        Sports      8.92      8.03      3.60         2.15          80.0   
14          Misc     15.00      4.89      0.24         1.69          61.0   
15        Sports      9.01      8.49      2.53         1.77          80.0   
16        Action      7.02      9.09      0.98         3.96          97.0   
17        Action      9.43      0.40      0.41        10.57          95.0   
18      Platform     12.78      3.75      3.54         0.55           NaN   
19          Misc      4.74      9.20      4.16         2.04          77.0   

   User_Score Rating  
0           8      E  
1         NaN    NaN  
2         8.3      E  
3           8      E  
4         NaN    NaN  
5         NaN    NaN  
6         8.5      E  
7         6.6      E  
8         8.4      E  
9         NaN    NaN  
10        NaN    NaN  
11        8.6      E  
12        NaN    NaN  
13        7.7      E  
14        6.3      E  
15        7.4      E  
16        8.2      M  
17          9      M  
18        NaN    NaN  
19        7.9      E

In [10]:
# Data cleaning- checking for any issues with the data 
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16715 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16713 non-null  object 
 1   Platform         16715 non-null  object 
 2   Year_of_Release  16446 non-null  float64
 3   Genre            16713 non-null  object 
 4   NA_sales         16715 non-null  float64
 5   EU_sales         16715 non-null  float64
 6   JP_sales         16715 non-null  float64
 7   Other_sales      16715 non-null  float64
 8   Critic_Score     8137 non-null   float64
 9   User_Score       10014 non-null  object 
 10  Rating           9949 non-null   object 
dtypes: float64(6), object(5)
memory usage: 1.4+ MB


### Issues found: 

- Lots of null values, and not all columns match the amount of null values
- Year of release should be a datetime object, currently is float 
- Critic score should be int, currently is float 
- User score should be float, is object 
<br>



## Data Cleaning

In [11]:
# Changing columns to lowercase and saving back to games 
games.columns = games.columns.str.lower()

games.head()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

   eu_sales  jp_sales  other_sales  critic_score user_score rating  
0     28.96      3.77         8.45          76.0          8      E  
1      3.58      6.81         0.77           NaN        NaN    NaN  
2     12.76      3.79         3.29          82.0        8.3      E  
3     10.93      3.28         2.95          80.0          8      E  
4      8.89     10.22         1.00           NaN        NaN    NaN

Dropping all null values from 'year_of_release' first because this column is extremely important to the analysis in this project. I'm also dropping them because there isn't another meaningful way to display this data alternatively if we proceed with the null values. 
<br>

- Why are the values missing? <br>
The values may be missing from the year_of_release because they are either so old or new that they may not have been processed correctly. We can find out more about this by reviewing all the other column data that shows up when we view a NaN year_of_release

<br> 
We will also cut out the rest of the null values for most of the columns to match the exact number of non- null values as year_of_release. The following columns will continue with null values that will be changes to a placeholder: 'critic_score', 'user_score' and 'rating'. These columns offer a bit more flexibility with continuing with null values.

In [12]:
# List of columns to clean by dropping null values
columns_to_clean = ['year_of_release', 'name', 'platform', 'genre', 'eu_sales', 'jp_sales', 'other_sales']

# Dropping null values from specified columns
games = games.dropna(subset=columns_to_clean)

games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16444 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16444 non-null  object 
 1   platform         16444 non-null  object 
 2   year_of_release  16444 non-null  float64
 3   genre            16444 non-null  object 
 4   na_sales         16444 non-null  float64
 5   eu_sales         16444 non-null  float64
 6   jp_sales         16444 non-null  float64
 7   other_sales      16444 non-null  float64
 8   critic_score     7983 non-null   float64
 9   user_score       9839 non-null   object 
 10  rating           9768 non-null   object 
dtypes: float64(6), object(5)
memory usage: 1.5+ MB


Looks good, all columns except for 'critic_score', 'user_score' and 'rating' now have 16444 non-null values. 

<br> 
Next, we'll move on to cleaning the column 'critic_score'
Then 'user_score' and finally, 'rating'

<br> 

- Why are there missing values?<br>
In these columns, the missing values have a more obvious reason- they just don't have a rating/ score yet. This could either be because the rating system started after the tracking of this data or because the rating has yet to processed or has never been completed. We can find more about these differences by reviewing just these columns with the year_of_release column to see if there are any patterns.

In [13]:
# Converting 'critic_score' to integer after handling missing values
games['critic_score'] = pd.to_numeric(games['critic_score'], errors='coerce').astype('Int64')

# Converting 'user_score' to float after handling missing values
games['user_score'] = pd.to_numeric(games['user_score'], errors='coerce')

# Filling missing values with placeholder -1 for ease of filtering later 
cols_to_fill = ['user_score', 'critic_score', 'rating']
games= games.fillna(value={'user_score': -1, 'critic_score': -1, 'rating': 'no rating'})

# Checking 
games.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16444 entries, 0 to 16714
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16444 non-null  object 
 1   platform         16444 non-null  object 
 2   year_of_release  16444 non-null  float64
 3   genre            16444 non-null  object 
 4   na_sales         16444 non-null  float64
 5   eu_sales         16444 non-null  float64
 6   jp_sales         16444 non-null  float64
 7   other_sales      16444 non-null  float64
 8   critic_score     16444 non-null  Int64  
 9   user_score       16444 non-null  float64
 10  rating           16444 non-null  object 
dtypes: Int64(1), float64(6), object(4)
memory usage: 1.5+ MB


In [14]:
# Checking that there are no outliers in the data for 'critic_score' and 'user_score'
# critic_score should have a max of 100 and user_score max of 10
games[['critic_score','user_score']].max()

critic_score    98.0
user_score       9.7
dtype: Float64

In [15]:
# Creating a new column for total_sales. This will combine sales from all regions and add a new column to the dataframe
games['total_sales'] = games[['na_sales', 'eu_sales', 'jp_sales', 'other_sales']].sum(axis=1)

games

name platform  year_of_release         genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16710  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16711               LMA Manager 2007     X360           2006.0        Sports   
16712        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16713               Spirits & Spells      GBA           2003.0      Platform   
16714            Winning Post 8 2016      PSV           2016.0    Simulation   

       na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score  \
0         41.36     28.96      3.77         8.45            76         8.0   
1         29.08      3.58      6.81         0.77            -1        -1.0   
2         15.68     12.76      3.79         3.29            82         8.3   
3         15.61     10.93      3.28         2.95            80         8.0   
4         11.27      8.89     10.22         1.00            -1        -1.0   
...         ...       ...       ...          ...           ...         ...   
16710      0.00      0.00      0.01         0.00            -1        -1.0   
16711      0.00      0.01      0.00         0.00            -1        -1.0   
16712      0.00      0.00      0.01         0.00            -1        -1.0   
16713      0.01      0.00      0.00         0.00            -1        -1.0   
16714      0.00      0.00      0.01         0.00            -1        -1.0   

          rating  total_sales  
0              E        82.54  
1      no rating        40.24  
2              E        35.52  
3              E        32.77  
4      no rating        31.38  
...          ...          ...  
16710  no rating         0.01  
16711  no rating         0.01  
16712  no rating         0.01  
16713  no rating         0.01  
16714  no rating         0.01  

[16444 rows x 12 columns]

In [16]:
# Checking 
games.head()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

   eu_sales  jp_sales  other_sales  critic_score  user_score     rating  \
0     28.96      3.77         8.45            76         8.0          E   
1      3.58      6.81         0.77            -1        -1.0  no rating   
2     12.76      3.79         3.29            82         8.3          E   
3     10.93      3.28         2.95            80         8.0          E   
4      8.89     10.22         1.00            -1        -1.0  no rating   

   total_sales  
0        82.54  
1        40.24  
2        35.52  
3        32.77  
4        31.38

The data has now been processed and is ready for analysis. 
<br>

## Data Analysis 